xbmc/guilib/GUIFontTTF.cpp

```diff
@@ -615,7 +633,102 @@ unsigned int CGUIFontTTF::GetTextureLineHeight() const
   return m_cellHeight + spacing_between_characters_in_texture;
 }
 
-CGUIFontTTF::Character* CGUIFontTTF::GetCharacter(character_t chr)
+std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecText& text)
+{
+  std::vector<Glyph> glyphs;
+  if (text.empty())
+  {
+    return glyphs;
+  }
+  std::vector<hb_script_t> scripts;
+  std::vector<RunInfo> runs;
+  hb_unicode_funcs_t* ufuncs = hb_unicode_funcs_create(hb_unicode_funcs_get_default());
+  hb_script_t lastScript;
+  int lastScriptIndex = -1;
+  int lastSetIndex = -1;
+
+  for (const auto& character : text)
+  {
+    scripts.emplace_back(hb_unicode_script(ufuncs, static_cast<wchar_t>(0xffff & character)));
+  }
+
+  // HB_SCRIPT_COMMON or HB_SCRIPT_INHERITED should be replaced with previous script
+  for (unsigned int i = 0; i < scripts.size(); i++)
+  {
+    if (scripts[i] == HB_SCRIPT_COMMON || scripts[i] == HB_SCRIPT_INHERITED)
+    {
+      if (lastScriptIndex != -1)
+      {
+        scripts[i] = lastScript;
+        lastSetIndex = i;
+      }
+    }
+    else
+    {
+      for (unsigned int j = lastSetIndex + 1; j < i; j++)
+        scripts[j] = scripts[i];
+      lastScript = scripts[i];
+      lastScriptIndex = i;
+      lastSetIndex = i;
+    }
+  }
+
+  lastScript = scripts[0];
+  int lastRunStart = 0;
+
+  for (unsigned int i = 0; i <= scripts.size(); i++)
+  {
+    if (i == scripts.size() || scripts[i] != lastScript)
+    {
+      RunInfo run{};
+      run.startOffset = lastRunStart;
+      run.endOffset = i;
+      run.script = lastScript;
+      runs.emplace_back(run);
+
+      if (i < scripts.size())
+      {
+        lastScript = scripts[i];
+        lastRunStart = i;
+      }
+      else
+      {
+        break;
+      }
+    }
+  }
+
+  for (auto& run : runs)
+  {
+    run.buffer = hb_buffer_create();
+    hb_buffer_set_direction(run.buffer, static_cast<hb_direction_t>(HB_DIRECTION_LTR));
+    hb_buffer_set_script(run.buffer, run.script);
```

Comment:
We also need to do have `hb_buffer_set_language` in order to render non-Chinese CJK text (~~C~~JK text? :wink:) correctly (see #19899 for a full explanation). I'm not sure if this code is used for rendering subtitles or not -- if it is used for rendering subtitles then you'd need to pass down the language code from the original caller because the behaviour for subtitle rendering and interface rendering should be:

 * Interface -- use the language code of the current interface language. I'm not sure if this is something that is explicitly provided by each language pack (though I expect it is), but you might need to change how language resources are loaded so that the language pack language is available here. 
 * Subtitles -- use the language code *of the subtitle* if available, otherwise fallback to the interface fallback.

Reply:
Yes this code is used for rendering subtitles and gui element texts. The caller passes down only `vecText` which is essentially vector of characters upper 32 bit as  colors/text and lower 32 bit as character.

Reply:
Passing down the language of text need a significant rework I think. 
How apps like browsers determine language while dealing with multilingual texts? Any idea?

Reply:
Kodi's current behaviour is  rendering  text character by character and no shaping is done. This makes text in Indic languages unreadable in Kodi. This PR only addresses that shaping issue currently.

Reply:
> Yes this code is used for rendering subtitles and gui element texts. The caller passes down only vecText which is essentially vector of characters upper 32 bit as colors/text and lower 32 bit as character.

Yeah so it seems like it'd need some rework (in the future -- I'm not suggesting you need to do it now) to handle text outside of the BMP. Good to know.

> How apps like browsers determine language while dealing with multilingual texts? Any idea?

Browsers use the language of the page (which is almost never set correctly), then they use the fallback list specified by the user in preferences (you can specify what languages you prefer in your browser settings), and then they have a hardcoded set of fallback fonts. The fallbacks I mentioned above are probably closest in spirit to how browsers deal with this problem, unless we want to add some extra configuration option so users can specify what CJK fallbacks they like.

Another option would be to auto-detect the language if the language is not explicitly set (for subtitles the subtitle track, for the interface the interface language) which is doable but would also be quite fragile (imagine if you have foreign text quoted within a subtitle, or there's some transcription error in a subtitle line) and so probably should be considered separately.

> Kodi's current behaviour is rendering text character by character and no shaping is done. This makes text in Indic languages unreadable in Kodi. This PR only addresses that shaping issue currently.

Sure, I just wanted to point out that Kodi also currently incorrectly renders Japanese and Korean text at the moment and the solution is not logically that complicated once Kodi switches to harfbuzz (though I agree it will probably require a bit of work to do -- if you like I can work on this separately).

Reply:
>  (though I agree it will probably require a bit of work to do -- if you like I can work on this separately).

Thats  great. And also dont hesitate to rework on this PR if required as I know the change in this PR  is a little bit ugly ( I'm very much new to kodi /harfbuzz / rendering and even to C++ too :) )

Reply:
I looked at this a bit more in the past few days and it seems to me that having proper CJK support would require several other bits of Kodi to be reworked (mainly font loading, as well as packaging a more complete font like Noto Sans CJK -- the Arial font pacakged with Kodi only contains Chinese glyphs it seems).

I will work on this in the future, but this shouldn't delay this PR being merged.